machine/v1beta1/types_vsphereprovider.go

```diff
@@ -87,12 +87,82 @@ type NetworkSpec struct {
+}
+
 // NetworkDeviceSpec defines the network configuration for a virtual machine's
 // network device.
 type NetworkDeviceSpec struct {
-	// NetworkName is the name of the vSphere network to which the device
-	// will be connected.
-	NetworkName string `json:"networkName"`
+	// networkName is the name of the vSphere network or port group to which the network
+	// device will be connected, for example, port-group-1. When not provided, the vCenter
+	// API will attempt to select a default network.
+	// The available networks (port groups) can be listed using `govc ls 'network/*'`
+	// +kubebuilder:example=port-group-1
+	// +kubebuilder:validation:MaxLength=80
+	// +optional
+	NetworkName string `json:"networkName,omitempty"`
+
+	// gateway is an IPv4 or IPv6 address which represents the subnet gateway,
+	// for example, 192.168.1.1.
+	// +kubebuilder:validation:Format=ipv4
+	// +kubebuilder:validation:Format=ipv6
+	// +kubebuilder:example=192.168.1.1
+	// +kubebuilder:example=2001:DB8:0000:0000:244:17FF:FEB6:D37D
+	// +optional
+	Gateway string `json:"gateway,omitempty"`
+
+	// ipAddrs is a list of one or more IPv4 and/or IPv6 addresses and CIDR to assign to
+	// this device, for example, 192.168.1.100/24. IP addresses provided via ipAddrs are
+	// intended to allow explicit assignment of a machine's IP address. IP pool configurations
+	// provided via addressesFromPool, however, defer IP address assignment to an external controller.
+	// If both addressesFromPool and ipAddrs are empty or not defined, DHCP will be used to assign
+	// an IP address. If both ipAddrs and addressesFromPools are defined, the IP addresses associated with
+	// both will be applied to the machine.
+	// +kubebuilder:validation:Format=ipv4
+	// +kubebuilder:validation:Format=ipv6
+	// +kubebuilder:example=192.168.1.100/24
+	// +kubebuilder:example=2001:DB8:0000:0000:244:17FF:FEB6:D37D/64
+	// +optional
+	IPAddrs []string `json:"ipAddrs,omitempty"`
+
+	// nameservers is a list of IPv4 and/or IPv6 addresses used as DNS nameservers, for example,
+	// 8.8.8.8. a nameserver is not provided by a fulfilled IPAddressClaim. If DHCP is not the
+	// source of IP addresses for this network device, nameservers should include a valid nameserver.
+	// +kubebuilder:validation:Format=ipv4
+	// +kubebuilder:validation:Format=ipv6
+	// +kubebuilder:example=8.8.8.8
+	// +optional
+	Nameservers []string `json:"nameservers,omitempty"`
+
+	// addressesFromPools is a list of references to IP pool types and instances which are handled
+	// by an external controller. addressesFromPool configurations provided via addressesFromPools
+	// defer IP address assignment to an external controller. IP addresses provided via ipAddrs,
+	// however, are intended to allow explicit assignment of a machine's IP address. If both
+	// addressesFromPool and ipAddrs are empty or not defined, DHCP will assign an IP address.
+	// If both ipAddrs and addressesFromPools are defined, the IP addresses associated with
```

Comment:
This looks good, only question I'm left with is which takes precedence, lets make sure it explains that IPAddrs come first (it might matter?)

Reply:
In upstream(and in the current incantation of the [mapi PR for this](https://github.com/openshift/machine-api-operator/pull/1079)) Neither take precedent, actually.  In the even that both are defined, multiple IP addresses are assigned to the adapter.

Reply:
I believe both IPs will get assigned to the same interface(in this case ens192) but I will test that out to be sure.  We are providing the IP addresses to the virtual machine via the network kargs.  I should know shortly.

Reply:
forgot to push the changes, they are up now.